src/lseq.rs

```diff
@@ -0,0 +1,269 @@
+//! # LSEQ
+//!
+//! An LSEQ tree is a CRDT for storing sequences of data (Strings, ordered lists).
+//! It provides an efficient view of the stored sequence, with fast index, insertion and deletion
+//! operations.
+//!
+//! LSEQ [1] is a member of the LOGOOT [2] family of algorithms for CRDT sequences. The major difference
+//! with LOGOOT is in the _allocation strategy_ that LSEQ employs to handle insertions.
+//!
+//! Internally, LSEQ views the sequence as the leaves of an ordered, exponential tree. An
+//! exponential tree is a tree where the number of childen grows exponentially with the depth of a
+//! node. For LSEQ, each layer of the tree doubles the available space. Each child is numbered from
+//! 0..2^(3+depth). The first and last child of a node cannot be turned into leaves.
+//!
+//! The path from the root of a tree to a leaf is called the _identifier_ of an element.
+//!
+//! The major challenge for LSEQs is the question of generating new identifiers for insertions.
+//!
+//! If we have the sequence of ordered pairs of identifiers and values `[ ix1: a , ix2: b , ix3: c ]`,
+//! and we want to insert `d` at the second position, we must find an identifer ix4 such that
+//! ix1 < ix4 < ix2. This ensures that every site will insert d in the same relative position in
+//! the sequence even if they dont have ix2 or ix1 yet. The [`IdentGen`] encapsulates this identifier
+//! generation, and ensures that the result is always between the two provided bounds.
+//!
+//! LSEQ is a CmRDT, to guarantee convergence it must see every operation. It also requires that
+//! they are delivered in a _causal_ order. Every deletion _must_ be applied _after_ it's
+//! corresponding insertion. To guarantee this property, use a causality barrier.
+//!
+//! [1] B. Nédelec, P. Molli, A. Mostefaoui, and E. Desmontils,
+//! “LSEQ: an adaptive structure for sequences in distributed collaborative editing,”
+//! in Proceedings of the 2013 ACM symposium on Document engineering - DocEng ’13,
+//! Florence, Italy, 2013, p. 37, doi: 10.1145/2494266.2494278.
+//!
+//! [2] S. Weiss, P. Urso, and P. Molli,
+//! “Logoot: A Scalable Optimistic Replication Algorithm for Collaborative Editing on P2P Networks,”
+//! in 2009 29th IEEE International Conference on Distributed Computing Systems,
+//! Montreal, Quebec, Canada, Jun. 2009, pp. 404–412, doi: 10.1109/ICDCS.2009.75.
+
+/// Contains the implementation of the exponential tree for LSeq
+pub mod ident;
+
+use ident::*;
+use serde::{Deserialize, Serialize};
+
+use crate::traits::CmRDT;
+
+use crate::vclock::{Actor, Dot};
+
+// SiteId can be generalized to any A if there is a way to generate a single invalid actor id at every site
+// Currently we rely on every site using the Id 0 for that purpose.
+
+/// Actor type for LSEQ sites.
+#[derive(Debug, Hash, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct SiteId(u32);
```

Comment:
What do you think of making this safer with an enum.
```rust
pub enum SiteId {
    Invalid,
    Valid(u32)
}
```
`derive(Ord)` on this should preserve `Invalid < Valid(...)`?

Reply:
I feel like we could actually remove `SiteId` entirely and leave it entirely abstract.

Reply:
There's still the point about safety around no actor being assigned the `A::default()` identifier. The other choice would be to use `Option<A>` anywhere a sentinel is needed

Reply:
That's what's currently done actually! Check the implementation of `ident.rs` the `Identifier` type uses an `Option<A>` everywhere.

Reply:
Ah very nice! (still slowly working my way through this as I find time in the day)

Reply:
`SiteId` probably doesn't make sense anymore now that it's `Option<A>` now right?

Reply:
Yep, lets remove it, we can consider a newtype for Actor separately from this work

Reply:
@xldenis mind making this change?